internals/plan/plan.go

```diff
@@ -1021,82 +1022,111 @@ func validServiceAction(action ServiceAction, additionalValid ...ServiceAction)
 		return nil, fmt.Errorf("cannot read layers directory: %v", err)
 	}
 
 	orders := make(map[int]string)
-	labels := make(map[string]int)
+	labels := make(map[string]string)
+	// Copy already loaded layer labels with full path to the layer file
+	// as they have come from a different import path.
+	for k, v := range lr.labels {
+		labels[k] = v
+	}
 
 	// Documentation says ReadDir result is already sorted by name.
 	// This is fundamental here so if reading changes make sure the
 	// sorting is preserved.
 	var layers []*Layer
 	for _, finfo := range finfos {
-		if finfo.IsDir() || !strings.HasSuffix(finfo.Name(), ".yaml") {
+		fileName := finfo.Name()
+		if finfo.IsDir() || !strings.HasSuffix(fileName, ".yaml") {
 			continue
 		}
 		// TODO Consider enforcing permissions and ownership here to
 		//      avoid mistakes that could lead to hacks.
-		match := fnameExp.FindStringSubmatch(finfo.Name())
+		match := fnameExp.FindStringSubmatch(fileName)
 		if match == nil {
-			return nil, fmt.Errorf("invalid layer filename: %q (must look like \"123-some-label.yaml\")", finfo.Name())
+			return nil, fmt.Errorf("invalid layer filename: %q (must look like \"123-some-label.yaml\")", fileName)
 		}
 
-		data, err := ioutil.ReadFile(filepath.Join(dirname, finfo.Name()))
+		filePath := filepath.Join(dirname, fileName)
+		data, err := ioutil.ReadFile(filePath)
 		if err != nil {
 			// Errors from package os generally include the path.
 			return nil, fmt.Errorf("cannot read layer file: %v", err)
 		}
 		label := match[2]
-		order, err := strconv.Atoi(match[1])
+		localOrder, err := strconv.Atoi(match[1])
 		if err != nil {
 			panic(fmt.Sprintf("internal error: filename regexp is wrong: %v", err))
 		}
 
-		oldLabel, dupOrder := orders[order]
-		oldOrder, dupLabel := labels[label]
-		if dupOrder {
-			oldOrder = order
-		} else if dupLabel {
-			oldLabel = label
+		if prevFile, dupOrder := orders[localOrder]; dupOrder {
+			return nil, fmt.Errorf("invalid layer filename: %q not unique (have %q already with same order %d)",
+				fileName, prevFile, localOrder)
 		}
-		if dupOrder || dupLabel {
-			return nil, fmt.Errorf("invalid layer filename: %q not unique (have \"%03d-%s.yaml\" already)", finfo.Name(), oldOrder, oldLabel)
+		if prevFile, dupLabel := labels[label]; dupLabel {
+			return nil, fmt.Errorf("invalid layer filename: %q not unique (have %q already with same label %q)",
+				fileName, prevFile, label)
 		}
+		orders[localOrder] = fileName
+		labels[label] = fileName
+		lr.labels[label] = filePath
 
-		orders[order] = label
-		labels[label] = order
-
-		layer, err := ParseLayer(order, label, data)
+		// globalOrder of the layers must be maintained when using import paths.
+		globalOrder := lr.offset + localOrder
+		layer, err := ParseLayer(globalOrder, label, data)
 		if err != nil {
 			return nil, err
 		}
+		layer.Source = filePath
 		layers = append(layers, layer)
 	}
+	if len(layers) > 0 {
+		lr.offset = layers[len(layers)-1].Order + 1
```

Comment:
Could just update this in the loop as we go? Would avoid the `if len(layers) > 0`.